### ISSUE TITLE
Evaluate manifold (ManifoldCAD) library for triangle mesh booleans

### ISSUE THREAD

**Issue by phkahler** [solvespace/solvespace#1371]
We still have a number of triangle mesh boolean bugs and meshes are also much slower than NURBS operations these days. I want to avoid CGAL because it's huge and we are small and simple. [The manifold library](https://github.com/elalish/manifold) looks promising.

All of our triangle meshes are initially created [here](https://github.com/solvespace/solvespace/blob/2e2aceb5eddb2dfd77296ce0325ddf4cfe836885/src/srf/shell.cpp#L588). It's a small function but there is a LOT of code under that. I would initially keep our NURBS triangulation code since it seems to work flawlessly. The other place we create meshes is when booleans are done.  We later do analysis on meshes for rendering, like edge classification and I'm not sure how easy that would be  - we could initially convert back to our own triangle mesh structure for that. The goal here would be to evaluate that library for correctness and see if it just squashes all of our remaining mesh bugs.

Also just want to ping the author @elalish and ask a few questions:
How small could we get your code for just triangle mesh booleans?
With and without GPU acceleration?
Your polygon triangulation is probably faster, but has no notion of curvature right?
Not sure if you're familiar with solvespace, does this seem like a good fit?

Edit: we could just add the other mesh type to our mesh objects. Create it if needed prior to a boolean and then convert the result back. That should allow the minimum work to test the booleans.

**Comment by phkahler** [comment]
> I'm not sure what kind of edge classification you're doing

It's used to highlight edges, either at face boundaries or silhouette edges. In the attached picture, the cube faces have white edges and the torus has silhouette edges which are view dependent. The view dependence seems to be handled quickly, but the setup after geometry changes is one of our slower operations - because our meshes are just a triangle soup. FaceID is important because we use it to pick faces which can have many triangles.

![Cube_Torus](https://github.com/solvespace/solvespace/assets/14852918/9e8db5ce-1f1b-460d-b64d-16a2642dc520)

**Comment by elalish** [comment]
Yes, the nice thing about a manifold mesh like ours is that it's easy to construct a halfedge data structure from it, so then any edge between triangles with different `originalID`s would be an edge to highlight. We use a halfedge structure internally - we might even consider exposing it if it's useful to enough people.
